data/src/test/java/org/apache/iceberg/io/TestTaskEqualityDeltaWriter.java

```diff
@@ -409,6 +421,55 @@ public void testUpsertDataWithFullRowSchema() throws IOException {
         .isEqualTo(ImmutableList.of(posRecord.copy("file_path", dataFile.path(), "pos", 0L)));
   }
 
+  @TestTemplate
+  public void testDeleteFileGranularity() throws IOException {
+    withGranularity(DeleteGranularity.FILE);
+  }
+
+  @TestTemplate
+  public void testDeletePartitionGranularity() throws IOException {
+    withGranularity(DeleteGranularity.PARTITION);
+  }
+
+  private void withGranularity(DeleteGranularity granularity) throws IOException {
+    List<Integer> eqDeleteFieldIds = Lists.newArrayList(idFieldId, dataFieldId);
+    Schema eqDeleteRowSchema = table.schema();
+
+    GenericTaskDeltaWriter deltaWriter =
+        createTaskWriter(eqDeleteFieldIds, eqDeleteRowSchema, granularity);
+
+    Map<Integer, Record> expected = Maps.newHashMapWithExpectedSize(2000);
+    // Create enough records, so we have multiple files
+    for (int i = 0; i < 2000; ++i) {
+      Record record = createRecord(i, "aaa" + i);
+      deltaWriter.write(record);
+      if (i % 5 == 10) {
+        deltaWriter.delete(record);
+      } else {
+        expected.put(i, record);
+      }
+    }
+
+    // Add some deletes in the end
+    for (int i = 0; i < 199; ++i) {
+      int id = i * 10 + 1;
+      Record record = createRecord(id, "aaa" + id);
+      deltaWriter.delete(record);
+      expected.remove(id);
+    }
+
+    WriteResult result = deltaWriter.complete();
+    assertThat(result.dataFiles()).as("Should have 2 data files.").hasSize(2);
+    assertThat(result.deleteFiles())
+        .as("Should have correct number of pos-delete files")
+        .hasSize(granularity.equals(DeleteGranularity.FILE) ? 2 : 1);
```

Comment:
should we also add assertion on the number of rows in the position delete files?